manPages/lilo.8
'\" t
.\" @(#)lilo.8 1.0 950728 aeb
.\" This page is based on the lilo docs, which carry the following
.\" COPYING condition:
.\"
.\" LILO program code, documentation and auxiliary programs are
.\" Copyright 1992-1994 Werner Almesberger.
.\" All rights reserved.
.\" 
.\" Redistribution and use in source and binary forms of parts of or the
.\" whole original or derived work are permitted provided that the
.\" original work is properly attributed to the author. The name of the
.\" author may not be used to endorse or promote products derived from
.\" this software without specific prior written permission. This work
.\" is provided "as is" and without any express or implied warranties.
.\"
.\" Original version, Andries Brouwer (aeb@cwi.nl), 950728
.\" Added t directive, as Daniel Quinlan asked, 950824
.\" Added L switch, John Coffman, 991118
.\"
.TH LILO 8 "20 March 2000"
.SH NAME
lilo \- install boot loader
.SH SYNOPSIS
Main function:
.LP
.B " /sbin/lilo"
\- install boot loader
.LP
Auxiliary uses:
.LP
.B " /sbin/lilo -q"
\- query map
.br
.B " /sbin/lilo -R"
\- set default command line for next reboot
.br
.B " /sbin/lilo -I"
\- inquire path name of current kernel
.br
.B " /sbin/lilo {-u|-U}"
\- uninstall lilo
.SH DESCRIPTION
.LP
.B lilo
installs a boot loader that will be activated next time you boot.
It has lots of options.
.LP
.TP
.B \-\^v
Increase verbosity. Giving one or more \fB-v\fP options
will make lilo more verbose.
.TP
.B \-\^q
List the currently mapped files.
.B lilo
maintains a file, by default
.IR "/boot/map" ,
containing the name and location of the kernel(s) to boot.
This option will list the names therein.
.TP
.BI "\-\^m " map-file
Use specified map file instead of the default.
.TP
.BI "\-\^C " config-file
.B lilo
reads its instructions about what files to map from its
config file, by default
.IR "/etc/lilo.conf" .
This option can be used to specify a non-default config file.
.TP
.BI "\-\^d " delay
If you have specified several kernels, and press Shift at boot-time,
the boot loader will present you with a choice of which system
to boot. After a timeout period the first kernel in the list
is booted. This option specifies the timeout delay in deciseconds.
.TP
.BI "\-\^D " label
Use the kernel with the given label, instead of the first one
in the list, as the default kernel to boot.
.TP
.BI "\-\^r " root-directory
Before doing anything else, do a \fIchroot\fP to the indicated
directory. Used for repairing a setup from a boot floppy.
.TP
.BI "\-\^t "
Test only. Do not really write a new boot sector or map file.
Use together with \fB-v\fP to find out what
.B lilo
is about to do.
.TP
.B "\-\^c"
Enable map compaction. This will merge read requests from adjacent
sectors. Speeds up the booting (especially from floppy).
.TP
.BI "\-\^f " disk-tab
Specify disk geometry parameter file. (The default is
.IR /etc/disktab .)
.TP
.BI "\-\^i " boot-sector
Specify a file to be used as the new boot sector. (The default is
.IR /boot/boot.b .)
.TP
.BI "\-\^l"
Generate linear sector addresses instead of sector/head/cylinder
addresses.
.TP
.BI "\-\^L"
Generate 32-bit Logical Block Addresses instead of C:H:S addresses, allowing
access to all partitions on disks greater than 8.4Gb.
.TP
.BI "\-\^P " "{fix|ignore}"
Fix (or ignore) `corrupt' partition tables, i.e., partition tables
with linear and sector/head/cylinder addresses that do not correspond.
.TP
.BI "\-\^s " save-file
When
.B lilo
overwrites the boot sector, it preserves the old contents in
a file, by default
.I /boot/boot.NNNN
where NNNN depends on the device. This option specifies an
alternate save file for the boot sector. (Or, together with the
\fB-u\fP option, specifies from where to restore the boot sector.)
.TP
.BI "\-\^S " save-file
Normally,
.B lilo
will not overwrite an existing save file. This options says
that overwriting is allowed.
.TP
.BI "\-\^u " device-name
Uninstall
.BR lilo ,
by copying the saved boot sector back. A time-stamp is checked.
.TP
.BI "\-\^U " device-name
Idem, but do not check the time-stamp.
.TP
.BI "\-\^R " "command line"
This option sets the default command for the boot loader the next
time it executes. The boot loader will then erase this line: this
is a once-only command. It is typically used in reboot scripts,
just before calling `shutdown -r'.
.TP
.BI "\-\^I " "label"
The label of the running kernel can be found in the environment
variable BOOT_IMAGE after startup. This command will print the
corresponding path name on stdout.
.TP
.B "\-\^V"
Print version number.

.LP
The above command line options correspond to the key words
in the config file indicated below.
.IP
.TS
l l.
-b bootdev	boot=bootdev
-c	compact
-d dsec	delay=dsec
-D label	default=label
-i bootsector	install=bootsector
-f file	disktab=file
-l	linear
-L	lba32
-m mapfile	map=mapfile
-P fix	fix-table
-P ignore	ignore-table
-s file	backup=file
-S file	force-backup=file
-v	verbose=level
.TE
.SH "SEE ALSO"
lilo.conf(5).
.br
The lilo distribution comes with very extensive documentation.
.SH "AUTHOR"
Werner Almesberger (almesber@bernina.ethz.ch).
.br
John Coffman (johninsd@san.rr.com) (lba32 extensions).


CHANGES

Changes from version 21.4.3 to 21.4.4 (10-Jun-2000) John Coffman

  Map Installer
  -------------
    - Added a minor RAID patch from Richard Bollinger: 'lilo-21_4_3.also'
      http://www.elliott-turbo.com/lilo/
    - Expand COMPACT/LINEAR warning message to include COMPACT/LBA32

  Boot Loader
  -----------
    - LARGE_EBDA option now loads second stage loader at 800:0000, and
      position of command line is changed; makes room for 512K kernels
      to be loaded low.
    - Minor optimizations when ES is reloaded (PUSH/POP) vs (mov thru AX)
    - Serial line now turns on DTR/RTS at initialization
    - Minor bugfix in numeric conversion (k/K-suffix)

  DISK.COM utility
  ----------------
    - The disk.com utility no longer requires 'nasm' to assemble.  The
      source code now makes the diagnostic floppy with "make floppy".


Changes from version 21.4.2 to 21.4.3 (06-May-2000) John Coffman

  Map Installer
  -------------
    - Added RAID patches from 'lilo.raid1.also' from Richard Bollinger
      http://www.elliott-turbo.com/lilo/lilo.raid1.also


Changes from version 21-3 to 21.4.2 (12-Apr-2000) -- John Coffman

  Boot loader
  -----------
    - bugfix in 'read.S' that disabled the combination 'linear' with
      'compact' ( '-l' and '-c' together).
    - fixed numeric convesions; kernel accepts octal, decimal, and hex, so 
      octal is now recognized by the boot loader.
    - bugfix patch where ES segment prefix was ommitted (RHL 'second.patch')
    - bugfix for compilation with 'LARGE_EBDA' option which caused command
      line passed to the kernel to be truncated at 78 characters
    
  Map installer
  -------------
    - RHL6.2 loopdev updates added
      
      
Changes from version 21 to 21-3 (24-FEB-2000)  -- John Coffman 
-----------------------------                  <johninsd@san.rr.com>

  Boot loader
  -----------
  
    - will boot from disks up to 2048Gb in size, as long as the BIOS allows
      the Enhanced BIOS calls for 32-bit sector addressing using packet calls.
    - checks if BIOS supports packet calls (int 0x13, AH=0x42), and uses
      these calls if 'lba32' was specified.  Otherwise, it uses
      the C:H:S addressing scheme of the original IBM-PC BIOS.
    - the disk read routine is separated into its own source file, "read.S",
      which is included in the first-stage, and second-stage loader codes.
      The first stage loader uses a very stripped down read routine, since it
      only deals in single sector transfers.
    - if BIOS does not support 32-bit packet calls, and the cylinder address
      overflows when the 32-bit address is converted to C:H:S, then a "9f"
      error code is issued.

  Map installer
  -------------
  
    - added the '-L' command line switch and 'lba32' configuration file 
      global option.  
    - map compaction of 'lba32' addresses is supported with the existing
      '-c' switch (or 'compact' configuration file global option).
    - floppy disks now will always fall back to C:H:S addressing.
    - changed the -v option to now accept a single digit to set the verbosity
      level; e.g., 'lilo -v 5 ...' replaces 'lilo -v -v -v -v -v ...'
      for maximum diagnostic output.
    - LILO version information is moved to the file 'version.h', and VERSION
      is now deleted.  This allows for tracking of VERSION_MAJOR (=21) and
      VERSION_MINOR (=3) for this and future code paths.
      
  BIOS tester
  -----------
  
    - Inspired by 'dparam.com', a much more extensive BIOS test program is
      included with this version.  Called 'disk.com', it was developed 
      under DOS, and requires 'nasm' to be assembled.
    - DISK.COM will probe the BIOS for disks supported by the BIOS, and report
      on their CHS geometry.  If EDD is supported, the EDD geometry is also
      reported.  If disks with cylinder numbers of 1024 or above are detected,
      it will issue a recommendation to use LILO with the 'lba32' option.
    - Creation of the bootable floppy image version of this utility is in-
      complete at this time.
      
    
Changes from version 20 to 21 (5-DEC-1998)
-----------------------------

  Boot loader
  -----------

    - cylinder calculation could overflow when using LINEAR (fix by Jonas
      Aaberg and an anonymous contributor)
    - first-stage loader code begins with the byte 0xFA (a CLI instruction).
      This works around the "INACCESSIBLE_BOOT_DEVICE" blue screen bug of
      Windows NT 4 (reported by too many people to mention)
    - new build-time option LARGE_EBDA to load boot loader in the 0x8xxxx area
      instead of the 0x9xxxx area (disabled by default; problem reported by
      Josef Moellers)

  Map installer
  -------------

    - new build-time option USE_TMPDIR to honor TMPDIR when creating temporary
      devices (disabled by default)
    - now only 0x1b6 bytes of the boot sector are overwritten, which preserves
      the passwords DR DOS stores there (fixed by Roy Millar)
    - -u and -U accidently implied -v
    - "BIOS drive 0xNN may not be accessible" warning was printed in cases
       where the drive code was never used (fixed by Andries Brouwer)
    - change rule for DOS16_big incorrectly used DOS16_small_hidden when
      hiding (fixed by Anthony Richardson)
    - RESET in CHANGE-RULES did not remove default rules (fixed by Kurt
      Garloff)
    - added support for Mylex DAC960 PCI RAID controllers (by Leonard Zubkoff)
    - added support for third to sixth IDE interface
    - build-time option READONLY was not used properly
    - fixed a few harmless compiler warnings

  Documentation and installation
  ------------------------------

    - user's guide claimed that partitions are hidden by setting the highest
      bit in the type (i.e. 0x80). The correct value is 0x10, which is also
      what is used in the code.

lilo.c
/* lilo.c  -  LILO command-line parameter processing */

/* Copyright 1992-1998 Werner Almesberger. See file COPYING for details. */


#include <stdlib.h>
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <fcntl.h>
#include <errno.h>
#include <sys/stat.h>

#include <asm/page.h>

#include "config.h"
#include "common.h"
#include "lilo.h"
#include "temp.h"
#include "device.h"
#include "geometry.h"
#include "map.h"
#include "bsect.h"
#include "cfg.h"
#include "identify.h"
#include "partition.h"
#include "md-int.h"

#define S2(x) #x
#define S(x) S2(x)
#define MAX_TOKEN 200

static void show_other(int fd)
{
    BOOT_SECTOR buf[SETUPSECS-1];
    const unsigned char *drvmap;
    const unsigned char *prtmap;

    if (read(fd,buf,sizeof(buf)) != sizeof(buf))
	die("Read on map file failed (access conflict ?)");
    if (!strncmp(buf[0].par_c.signature-4,"LILO",4)) {
	printf("    Pre-21 signature (0x%02x,0x%02x,0x%02x,0x%02x)\n",
	  buf[0].par_c.signature[0],buf[0].par_c.signature[1],
	  buf[0].par_c.signature[2],buf[0].par_c.signature[3]);
	return;
    }
    if (strncmp(buf[0].par_c.signature,"LILO",4)) {
	printf("    Bad signature (0x%02x,0x%02x,0x%02x,0x%02x)\n",
	  buf[0].par_c.signature[0],buf[0].par_c.signature[1],
	  buf[0].par_c.signature[2],buf[0].par_c.signature[3]);
	return;
    }
    drvmap = ((unsigned char *) buf+buf[0].par_c.drvmap);
    prtmap = drvmap+2*(DRVMAP_SIZE+1);
    while (drvmap[0] && drvmap[1]) {
	printf("    BIOS drive 0x%02x is mapped to 0x%02x\n",drvmap[0],
	  drvmap[1]);
	drvmap += 2;
    }
    while (prtmap[0] && prtmap[1]) {
	printf("    BIOS drive 0x%02x, offset 0x%x: 0x%02x -> 0x%02x\n",
	  prtmap[0],prtmap[1]+PART_TABLE_OFFSET,prtmap[2],prtmap[3]);
	prtmap += 4;
    }
}


static void show_images(char *map_file)
{
    DESCR_SECTORS descrs;
    BOOT_SECTOR boot;
    GEOMETRY geo;
    SECTOR_ADDR addr[4];
    char buffer[SECTOR_SIZE];
    char *name;
    int fd,image,i;
    int tsecs;
    int tlinear, tlba32;
    unsigned short checksum,flags;

    fd = geo_open(&geo,map_file,O_RDONLY);
    if (read(fd,buffer,SECTOR_SIZE) != SECTOR_SIZE)
	die("read %s: %s",map_file,strerror(errno));
    if (read(fd,(char *) &descrs,sizeof(descrs)) != sizeof(descrs))
	die("read %s: %s",map_file,strerror(errno));
    tlinear = (descrs.d.descr[0].start.device & LINEAR_FLAG) != 0;
    tlba32  = (descrs.d.descr[0].start.device & LBA32_FLAG) != 0;
    if (tlinear != linear  ||  tlba32 != lba32) {
        printf("Warning: mapfile created with %s option\n",
	       tlinear?"linear":tlba32?"lba32":"no linear/lba32");
        linear = tlinear;  lba32 = tlba32;
    }
    if (verbose > 0) {
	bsect_read(cfg_get_strg(cf_options,"boot"),&boot);
	printf("Global settings:\n");
	tsecs = (boot.par_1.delay*55+50)/100;
	printf("  Delay before booting: %d.%d seconds\n",tsecs/10,tsecs % 10);
	if (boot.par_1.timeout == 0xffff) printf("  No command-line timeout\n");
	else {
	    tsecs = (boot.par_1.timeout*55+50)/100;
	    printf("  Command-line timeout: %d.%d seconds\n",tsecs/10,
	      tsecs % 10);
	}
	if (boot.par_1.prompt) printf("  Always enter boot prompt\n");
	else printf("  Enter boot prompt only on demand\n");
	if (!boot.par_1.port) printf("  Serial line access is disabled\n");
	else printf("  Boot prompt can be accessed from COM%d\n",
	      boot.par_1.port);
	if (!boot.par_1.msg_len) printf("  No message for boot prompt\n");
	else printf("  Boot prompt message is %d bytes\n",boot.par_1.msg_len);
	if (*(unsigned short *) buffer != DC_MAGIC || !buffer[2])
	    printf("  No default boot command line\n");
	else printf("  Default boot command line: \"%s\"\n",buffer+2);
	printf("Images:\n");
    }
    for (image = 0; image < MAX_IMAGES; image++)
	if (*(name = descrs.d.descr[image].name)) {
	    printf("%s%-" S(MAX_IMAGE_NAME) "s%s",verbose > 0 ? "  " : "",name,
	      image ? "  " : " *");
	    if (verbose > 1) {
	        if (descrs.d.descr[image].start.device & (LINEAR_FLAG|LBA32_FLAG)) {
		   unsigned int sector;
		   sector = descrs.d.descr[image].start.device & LBA32_NOCOUNT ?
		        descrs.d.descr[image].start.num_sect : 0;
		   sector = (sector<<8)+descrs.d.descr[image].start.head;
	           sector = (sector<<8)+descrs.d.descr[image].start.track;
		   sector = (sector<<8)+descrs.d.descr[image].start.sector;
		   printf(" <dev=0x%02x,linear=%d>",
		     descrs.d.descr[image].start.device,
		     sector);
		}
	        else { /*  CHS addressing */
		    printf(" <dev=0x%02x,hd=%d,cyl=%d,sct=%d>",
		      descrs.d.descr[image].start.device,
		      descrs.d.descr[image].start.head,
		      descrs.d.descr[image].start.track,
		      descrs.d.descr[image].start.sector);
		}
	    }
	    printf("\n");
	    if (verbose > 0) {
		flags = descrs.d.descr[image].flags;
		if (!*descrs.d.descr[image].password)
		    printf("    No password\n");
		else printf("    Password is required for %s\n",flags &
		      FLAG_RESTR ? "specifying options" : "booting this image");
		printf("    Boot command-line %s be locked\n",flags &
		  FLAG_LOCK ? "WILL" : "won't");
		printf("    %single-key activation\n",flags & FLAG_SINGLE ?
		  "S" : "No s");
		if (flags & FLAG_KERNEL) {
#ifdef NORMAL_VGA
		    if (!(flags & FLAG_VGA))
		       printf("    VGA mode is taken from boot image\n");
		    else {
			printf("    VGA mode: ");
			switch (descrs.d.descr[image].vga_mode) {
			    case NORMAL_VGA:
				printf("NORMAL\n");
				break;
			    case EXTENDED_VGA:
				printf("EXTENDED\n");
				break;
			    case ASK_VGA:
				printf("ASK\n");
				break;
			    default:
				printf("%d (0x%04x)\n",
				  descrs.d.descr[image].vga_mode,
				  descrs.d.descr[image].vga_mode);
			}
		    }
#endif
		    if (!descrs.d.descr[image].start_page)
			printf("    Kernel is loaded \"low\"\n");
		    else printf("    Kernel is loaded \"high\", at 0x%08lx\n",
			  (unsigned long) descrs.d.descr[image].start_page*
			  PAGE_SIZE);
		    if (!*(unsigned long *) descrs.d.descr[image].rd_size)
			printf("    No initial RAM disk\n");
		    else printf("    Initial RAM disk is %ld bytes\n",
			  *(unsigned long *) descrs.d.descr[image].rd_size);
		}
		if (!geo_find(&geo,descrs.d.descr[image].start)) {
		    printf("    Map sector not found\n");
		    continue;
		}
		if (read(fd,addr,4*sizeof(SECTOR_ADDR)) !=
		  4*sizeof(SECTOR_ADDR))
			die("Read on map file failed (access conflict ?)");
		if (!geo_find(&geo,addr[0]))
		    printf("    Fallback sector not found\n");
		else {
		    if (read(fd,buffer,SECTOR_SIZE) != SECTOR_SIZE)
			die("Read on map file failed (access conflict ?)");
		    if (*(unsigned short *) buffer != DC_MAGIC)
			printf("    No fallback\n");
		    else printf("    Fallback: \"%s\"\n",buffer+2);
		}
		if (flags & FLAG_KERNEL)
		    if (!geo_find(&geo,addr[1]))
			printf("    Options sector not found\n");
		    else {
			if (read(fd,buffer,SECTOR_SIZE) != SECTOR_SIZE)
			    die("Read on map file failed (access conflict ?)");
			printf("    Options: \"%s\"\n",buffer);
		    }
		else if (geo_find(&geo,addr[3])) show_other(fd);
		   else printf("    Image data not found\n");
	    }
	}
    (void) close(fd);
    checksum = INIT_CKS;
    for (i = 0; i < sizeof(descrs)/sizeof(unsigned short); i++)
	checksum ^= ((unsigned short *) &descrs)[i];
    if (!checksum) exit(0);
    fflush(stdout);
    fprintf(stderr,"Checksum error\n");
    exit(1);
}


static void usage(char *name)
{
    char *here;

    here = strrchr(name,'/');
    if (here) name = here+1;
    fprintf(stderr,"usage: %s [ -C config_file ] -q [ -m map_file ] "
      "[ -v ... ]\n",name);
    fprintf(stderr,"%7s%s [ -C config_file ] [ -b boot_device ] [ -c ] "
      "[ -l | -L ]\n","",name);
    fprintf(stderr,"%12s[ -i boot_sector ] [ -m map_file ] [ -d delay ]\n","");
    fprintf(stderr,"%12s[ -v ... ] [ -t ] [ -s save_file | -S save_file ]\n",
      "");
    fprintf(stderr,"%12s[ -P fix | -P ignore ] [ -r root_dir ] [ -w ]\n","");
    fprintf(stderr,"%7s%s [ -C config_file ] [ -m map_file ] "
      "-R [ word ... ]\n","",name);
    fprintf(stderr,"%7s%s [ -C config_file ] -I name [ options ]\n","",name);
    fprintf(stderr,"%7s%s [ -C config_file ] [ -s save_file ] "
      "-u | -U [ boot_device ]\n","",name);
    fprintf(stderr,"%7s%s -V\n\n","",name);
    exit(1);
}


int main(int argc,char **argv)
{
    char *name,*config_file,*reboot_arg,*identify,*ident_opt,*new_root;
    char *uninst_dev;
    int query,more,version,uninstall,validate,do_md_install,pass;
    BOOT_SECTOR dummy;
    IMAGE_DESCR dummy2;
    struct stat st;
    int fd, md_fd;
    md_array_info_t md_array_info;
    md_disk_info_t md_disk_info;
    char md_boot_name[MAX_TOKEN+1];
    char md_boot_map[MAX_TOKEN+1];
    DT_ENTRY md_disk;
    DT_ENTRY *disk;

    config_file = DFL_CONFIG;
    reboot_arg = identify = ident_opt = new_root = uninst_dev = NULL;
    pass = do_md_install = query = version = uninstall = validate = 0;
    name = *argv++;
    argc--;
    cfg_init(cf_options);
    while (argc && **argv == '-') {
	argc--;
	if (argv[0][2]) usage(name);
	switch ((*argv++)[1]) {
	    case 'b':
		if (!argc) usage(name);
		cfg_set(cf_options,"boot",*argv++,NULL);
		argc--;
		break;
	    case 'c':
		cfg_set(cf_options,"compact",NULL,NULL);
		compact = 1;
		break;
	    case 'd':
		if (!argc) usage(name);
		cfg_set(cf_options,"delay",*argv++,NULL);
		argc--;
		break;
	    case 'D':
		if (!argc) usage(name);
		cfg_set(cf_options,"default",*argv++,NULL);
		argc--;
		break;
	    case 'f':
		if (!argc) usage(name);
		cfg_set(cf_options,"disktab",*argv++,NULL);
		argc--;
		break;
	    case 'l':
		cfg_set(cf_options,"linear",NULL,NULL);
		linear = 1;
		break;
	    case 'L':
		cfg_set(cf_options,"lba32",NULL,NULL);
		lba32 = 1;
		break;
	    case 'm':
		if (!argc) usage(name);
		cfg_set(cf_options,"map",*argv++,NULL);
		argc--;
		break;
	    case 'i':
		if (!argc) usage(name);
		cfg_set(cf_options,"install",*argv++,NULL);
		argc--;
		break;
	    case 'I':
		if (!argc) usage(name);
		identify = *argv++;
		if (--argc) {
		    ident_opt = *argv++;
		    argc--;
		}
		break;
	    case 'X':
		printf("-DIMAGES=%d -DCODE_START_1=%d -DCODE_START_2=%d "
		  "-DDESCR_SIZE=%d -DDSC_OFF=%d -DDSC_OFF2=%d -DDFCMD_OFF=%d "
		  "-DMSG_OFF=%d -DFLAGS_OFF=%d\n",MAX_IMAGES,
		  sizeof(BOOT_PARAMS_1),sizeof(BOOT_PARAMS_2),
		  sizeof(IMAGE_DESCR),
		  (void *) &dummy.par_1.descr[0]-(void *) &dummy,
		  (void *) &dummy.par_1.descr[1]-(void *) &dummy,
		  (void *) &dummy.par_1.descr[2]-(void *) &dummy,
		  (void *) &dummy.par_1.msg_len-(void *) &dummy,
		  (void *) &dummy2.flags-(void *) &dummy2);
		exit(0);
	    case 'C':
		if (!argc) usage(name);
		config_file = *argv++;
		argc--;
		break;
	    case 'S':
		if (!argc) usage(name);
		cfg_set(cf_options,"force-backup",*argv++,NULL);
		argc--;
		break;
	    case 's':
		if (!argc) usage(name);
		cfg_set(cf_options,"backup",*argv++,NULL);
		argc--;
		break;
	    case 'P':
		if (!argc) usage(name);
		if (!strcmp(*argv,"fix"))
		    cfg_set(cf_options,"fix-table",NULL,NULL);
		else if (!strcmp(*argv,"ignore"))
			cfg_set(cf_options,"ignore-table",NULL,NULL);
		    else usage(name);
		argv++;
		argc--;
		break;
	    case 'q':
		query = 1;
		break;
	    case 'r':
		if (!argc) usage(name);
		new_root = *argv++;
		argc--;
		break;
	    case 'R':
		if (!argc) reboot_arg = "";
		else while (argc) {
			if (!reboot_arg)
			    *(reboot_arg = alloc(strlen(*argv)+1)) = 0;
			else strcat(reboot_arg = ralloc(reboot_arg,
			      strlen(reboot_arg)+strlen(*argv)+2)," ");
			strcat(reboot_arg,*argv++);
			argc--;
		    }
		break;
	    case 't':
		test = 1;
		break;
	    case 'u':
		validate = 1;
		/* fall through */
	    case 'U':
		uninstall = 1;
		if (argc) {
		    if (argc-- > 1) usage(name);
		    uninst_dev = *argv;
		}
		break;
	    case 'v':
	        if (argc && **argv>='0' && **argv<='9') {
		    verbose = **argv - '0';
		    argc--;
		    argv++;
		} else
	            verbose++;
		break;
	    case 'V':
		version = 1;
		break;
	    case 'w':
		cfg_set(cf_options,"nowarn",NULL,NULL);
		nowarn = 1;
		break;
	    default:
		usage(name);
	}
    }
    if (!new_root) new_root = getenv("ROOT");
    if (new_root && *new_root) {
	if (chroot(new_root) < 0) die("chroot %s: %s",new_root,strerror(errno));
	if (chdir("/") < 0) die("chdir /: %s",strerror(errno));
    }
    if (atexit(temp_remove)) die("atexit() failed");
    if (verbose > 0 || version) {
       printf("LILO version %d.%d%s%s", VERSION_MAJOR, VERSION_MINOR,
	      VERSION_EDIT, test ? " (test mode)" : "");
	if (version && !verbose) {
	    printf("\n");
	    return 0;
	}
	printf(", Copyright (C) 1992-1998 Werner Almesberger\n"
	       "'lba32' extensions Copyright (C) 1999,2000 John Coffman\n");
        if (verbose>=4) printf("Released %s and compiled at %s on %s.\n",
	        VERSION_DATE, __TIME__, __DATE__);
        printf("\n");
        if (version) return 0;
    }
#if 0
    if (((install || test || boot_device || disktab_file || compact) && !argc)
      || (compact && linear && 0)) usage(name);
#endif
    if (!nowarn && compact && (linear || lba32))
	fprintf(stderr,"Warning: COMPACT may conflict with %s on some "
	  "systems\n", lba32 ? "LBA32" : "LINEAR");
    preload_types();
    fd = cfg_open(config_file);
    more = cfg_parse(cf_options);
    if (!nowarn) {
	if (fstat(fd,&st) < 0) {
	    fprintf(stderr,"fstat %s: %s\n",config_file,strerror(errno));
	    exit(1);
	}
	if (S_ISREG(st.st_mode)) {
	    if (st.st_uid)
		fprintf(stderr,"Warning: %s should be owned by root\n",
		  config_file);
	    else if (st.st_mode & (S_IWGRP | S_IWOTH))
		    fprintf(stderr,"Warning: %s should be writable only for "
		      "root\n",config_file);
		else if ((cfg_get_strg(cf_all,"password") || cfg_get_strg(
		      cf_options,"password")) && (st.st_mode & (S_IRGRP |
		      S_IROTH)))
			fprintf(stderr,"Warning: %s should be readable only "
			  "for root if using PASSWORD\n",config_file);
	}
    }
    preload_dev_cache();
    if (identify) identify_image(identify,ident_opt);
    if (strncmp("/dev/md",cfg_get_strg(cf_options,"boot"),7) == 0) {
	if ((md_fd=open(cfg_get_strg(cf_options,"boot"),O_NOACCESS)) < 0)
	    die("Unable to open %s",cfg_get_strg(cf_options,"boot"));
	if (fstat(md_fd,&st) < 0)
	    die("Unable to stat %s",cfg_get_strg(cf_options,"boot"));
	if (!S_ISBLK(st.st_mode))
	    die("%s is not a block device",cfg_get_strg(cf_options,"boot"));
	if (ioctl(md_fd,GET_ARRAY_INFO,&md_array_info) < 0)
	    die("Unable to get RAID info on %s",cfg_get_strg(cf_options,"boot"));
	if ((md_array_info.major_version == 0) && (md_array_info.minor_version < 90))
	    die("Raid versions < 0.90 are not supported");
	if (md_array_info.level != 1)
	    die("Only RAID1 devices are supported as boot devices");
	do_md_install = 1;
	strcpy(md_boot_name,cfg_get_strg(cf_options,"boot"));
	if (cfg_get_strg(cf_options,"map"))
	    strcpy(md_boot_map,cfg_get_strg(cf_options,"map"));
	else
	    strcpy(md_boot_map,MAP_FILE);
	md_disk.device = (MD_MAJOR << 8) | md_array_info.md_minor;
	md_disk.bios = 0x80;
	md_disk.next = disktab;
	disktab = &md_disk;
    }
    while( (pass == 0) || (do_md_install && (pass < md_array_info.nr_disks)) ) {
	if(do_md_install) {
	    GEOMETRY geo;
	    DEVICE dev;
	    int device,disk_fd;
	    char new_name[MAX_TOKEN+1];

	    if(pass > 0) {
		close(fd);
		cfg_init(cf_options);
		fd = cfg_open(config_file);
		more = cfg_parse(cf_options);
	    }
	    md_disk_info.number = pass;
	    if (ioctl(md_fd,GET_DISK_INFO,&md_disk_info) < 0)
		die("main: GET_DISK_INFO: %s", strerror(errno));
	    device = (md_disk_info.major << 8) | md_disk_info.minor;
	    if (device == 0) { /* empty slot left over from recovery process */
		pass++;
		continue;
	    }
	    disk_fd = dev_open(&dev,device,O_NOACCESS);
	    if (md_disk_info.state & (1 << MD_DISK_FAULTY)) {
		printf("disk %s marked as faulty, skipping\n",dev.name);
		pass++;
		continue;
	    }
	    geo_query_dev(&geo,device,1);
	    disk = alloc_t(DT_ENTRY);
	    disk->bios = 0x80;
	    disk->device = device & 0xfff0;
	    disk->sectors = geo.sectors;
	    disk->heads = geo.heads;
	    disk->cylinders = geo.cylinders;
	    disk->start = geo.start;
	    disk->next = disktab;
	    disktab = disk;
	    if (cfg_get_strg(cf_options,"boot")) cfg_unset(cf_options,"boot");
	    if (cfg_get_strg(cf_options,"map")) cfg_unset(cf_options,"map");
	    strncpy(new_name,dev.name,8);
	    new_name[8] = '\0';
	    cfg_set(cf_options,"boot",new_name,NULL);
	    snprintf(new_name,MAX_TOKEN,"%s.%04x",md_boot_map,device);
	    cfg_set(cf_options,"map",new_name,NULL);
	    printf("boot = %s, map = %s\n", cfg_get_strg(cf_options,"boot"),
		cfg_get_strg(cf_options,"map"));
	    md_disk.sectors = geo.sectors;
	    md_disk.heads = geo.heads;
	    md_disk.cylinders = geo.cylinders;
	    md_disk.start = geo.start;
	}
	    
	pass++;
	if (uninstall)
	    bsect_uninstall(uninst_dev ? uninst_dev : cfg_get_strg(cf_options,
	      "boot"),cfg_get_strg(cf_options,"backup"),validate);
	compact = cfg_get_flag(cf_options,"compact");
	linear = cfg_get_flag(cf_options,"linear");
        lba32  = cfg_get_flag(cf_options,"lba32");
	nowarn = cfg_get_flag(cf_options,"nowarn");
	if (cfg_get_strg(cf_options,"verbose"))
	    verbose += to_number(cfg_get_strg(cf_options,"verbose"));
	if (reboot_arg) {
	    map_patch_first(cfg_get_strg(cf_options,"map") ? cfg_get_strg(
	      cf_options,"map") : MAP_FILE,reboot_arg);
	    exit(0);
	}
	if (argc) usage(name);
	geo_init(cfg_get_strg(cf_options,"disktab"));
	if (query)
	    show_images(!cfg_get_strg(cf_options,"map") ? MAP_FILE :
	      cfg_get_strg(cf_options,"map"));
	bsect_open(cfg_get_strg(cf_options,"boot"),cfg_get_strg(cf_options,"map") ?
	  cfg_get_strg(cf_options,"map") : MAP_FILE,cfg_get_strg(cf_options,
	  "install"),cfg_get_strg(cf_options,"delay") ? to_number(cfg_get_strg(
	  cf_options,"delay")) : 0,cfg_get_strg(cf_options,"timeout") ?
	  to_number(cfg_get_strg(cf_options,"timeout")) : -1);
	if (more) {
            cfg_init(cf_top);
            if (cfg_parse(cf_top)) cfg_error("Syntax error");
	}
	if (!bsect_number()) die("No images have been defined.");
	check_fallback();
	if (!test)
	    if (cfg_get_strg(cf_options,"force-backup"))
		bsect_update(cfg_get_strg(cf_options,"force-backup"),1);
	    else bsect_update(cfg_get_strg(cf_options,"backup"),0);
	else {
	    bsect_cancel();
	    fprintf(stderr,"The boot sector and the map file have *NOT* been "
	      "altered.\n");
	}
    }
    return 0;
}
